crates/storage/provider/src/providers/static_file/writer.rs

```diff
@@ -48,25 +51,28 @@ impl<'a> StaticFileProviderRW<'a> {
     fn open(
         segment: StaticFileSegment,
         block: u64,
-        reader: StaticFileProvider,
+        reader: Weak<StaticFileProviderInner>,
         metrics: Option<Arc<StaticFileProviderMetrics>>,
-    ) -> ProviderResult<(NippyJarWriter<'a, SegmentHeader>, PathBuf)> {
+    ) -> ProviderResult<(NippyJarWriter<SegmentHeader>, PathBuf)> {
         let start = Instant::now();
 
+        let static_file_provider = StaticFileProvider(reader.upgrade().unwrap());
```

Comment:
this needs .expects 

and a note that this is fine because this is only accessible as a RefMut, so the owned type always exists,
in other words it's not possible to detach the StaticFileProviderRW from the `StaticFileProvider`

Reply:
added